model/src/main/kotlin/DependencyGraph.kt

```diff
@@ -0,0 +1,240 @@
+ *
+ * To further reduce memory consumption, package identifiers are not repeated, but are listed once and then referenced
+ * by indices. The storage format defined by this class is not really human-readable, but it causes a significant
+ * reduction of memory usage. On deserialization, the information stored can be manually converted again to a structure
+ * with [Scope] and [PackageReference] objects (with shared references, so keeping memory consumption low, too), so
+ * that it can be processed in the usual ways.
+ */
+data class DependencyGraph(
+    /**
+     * Stores the dependency graph as a list of fragments. Each fragment can contain an arbitrary number of
+     * dependency trees, which are consistent in themselves.
+     */
+    val trees: Set<DependencyReference>,
+
+    /**
+     * A list with the identifiers of dependencies. This list is used to resolve the numeric indices contained in the
+     * [DependencyReference] objects.
+     */
+    val identifierList: List<String>,
+
+    /**
+     * A mapping from scope names to the direct dependencies of the scopes. Based on this information, the set of
+     * [Scope]s of a project can be constructed from the serialized form.
+     */
+    val scopeMapping: Map<String, List<DependencyIndex>>,
+
+    /** The name of the analyzer implementation which created this graph. */
+    val managerName: String
+) {
+    /**
+     * Transform the data stored in this object to the classical layout of dependency information, which is a set of
+     * [Scope]s referencing the packages they depend on.
+     */
+    fun createScopes(): SortedSet<Scope> {
+        val refMapping = constructReferenceMapping(managerName)
+
+        return scopeMapping.mapTo(sortedSetOf<Scope>()) { entry ->
+            val dependencies = entry.value.mapTo(sortedSetOf()) { index ->
+                refMapping[index.toKey()] ?: throw IllegalStateException("Could not resolve dependency index $index.")
+            }
+            Scope(entry.key, dependencies)
+        }
+    }
+
+    /**
+     * Construct a mapping from dependency indices to [PackageReference] objects using the given [managerName]. Based
+     * on this mapping, the structure with [Scope]s can be generated.
+     */
+    private fun constructReferenceMapping(managerName: String): Map<String, PackageReference> {
+        val refMapping = mutableMapOf<String, PackageReference>()
+        trees.forEach { tree ->
+            constructReferenceTree(managerName, tree, refMapping)
+        }
+
+        return refMapping
+    }
+
+    /**
+     * Construct the tree with [PackageReference]s by navigating the dependency tree starting with [ref] and
+     * populate the given [refMapping]. Use [managerName] to construct package identifiers.
+     */
+    private fun constructReferenceTree(
+        managerName: String,
+        ref: DependencyReference,
+        refMapping: MutableMap<String, PackageReference>
+    ):
+            PackageReference {
```

Comment:
That looks really odd. Let's do

    ): PackageReference {

Reply:
Obviously, one cannot rely on the IDE's formatting capabilities here.